programs/ssl/ssl_client2.c

```diff
@@ -3465,94 +3457,63 @@ int main( int argc, char *argv[] )
         ret = 0;
     }
 
+#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET)
     if( opt.reconnect != 0 )
     {
         mbedtls_printf("  . Saving session for reuse..." );
         fflush( stdout );
 
-        /* For TLS <= 1.2, serialize the session and restore later,
-         * triggering either session ID or session ticket based
-         * resumption in the next handshake. */
-#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
-        if( ssl.minor_ver == MBEDTLS_SSL_MINOR_VERSION_4 )
+        if( opt.reco_mode == 1 )
         {
-            /*
-             * Store ticket sent by the server after the handshake is completed.
-             */
-
-#if defined(MBEDTLS_SSL_NEW_SESSION_TICKET)
-            ret = mbedtls_ssl_get_client_ticket( &ssl, &ticket );
-            if( ret < 0 )
+            /* free any previously saved data */
+            if( session_data != NULL )
             {
-                mbedtls_printf( " failed\n  ! mbedtls_ssl_get_ticket returned -0x%x\n\n",
-                                (unsigned) -ret );
-                goto exit;
+                mbedtls_platform_zeroize( session_data, session_data_len );
+                mbedtls_free( session_data );
+                session_data = NULL;
             }
-            else if( ret == 0 )
+
+            /* get size of the buffer needed */
+            mbedtls_ssl_session_save( mbedtls_ssl_get_session_pointer( &ssl ),
```

Comment:
This isn't compatible with the intended ticket API, as far as I see: The ticket is retrieved as an instance of `mbedtls_ssl_session` via `mbedtls_ssl_get_session()`, and then serialized via `mbedtls_ssl_session_save()`. `mbedtls_ssl_get_session_pointer()` assumes that there's "the" session/ticket attacked to an SSL context, which isn't true anymore in TLS 1.3. 

We can change this at a later point, though.

Reply:
Let's fix this at a later stage.

Reply:
Ok